google-ads-examples/src/main/java/com/google/ads/googleads/examples/recommendations/optimizations/ApplyRecommendations.java

```diff
@@ -0,0 +1,445 @@
+      Map<String, AggregatedRecommendation> aggregatedRecommendations,
+      String reportDirectory)
+      throws IOException {
+    Path optiDir = Paths.get(reportDirectory, OPTIMIZATION_DIR_PREFIX + customer.getId());
+    if (!Files.exists(optiDir)) {
+      Files.createDirectories(optiDir);
+    }
+
+    // Generates 'optiScore.csv' for the optimization reports.
+    Path csvPath = optiDir.resolve(OPTI_OPTIMIZATION_SCORE_CSV);
+    try (CsvWriter csv = CsvWriter.builder().build(csvPath, StandardCharsets.UTF_8)) {
+      csv.writeRow(OPTI_OPTIMIZATION_SCORE_CSV_COLUMNS)
+          .writeRow(
+              String.valueOf(customer.getId()),
+              customer.getDescriptiveName(),
+              String.valueOf(customer.getOptimizationScore()),
+              String.valueOf(getCustomerOptiScore(googleAdsClient, customer.getId())));
+    }
+
+    // Generates 'recommendations.csv' for the optimization reports.
+    csvPath = optiDir.resolve(OPTI_RECOMMENDATIONS_CSV);
+    try (CsvWriter csv = CsvWriter.builder().build(csvPath, StandardCharsets.UTF_8)) {
+      csv.writeRow(OPTI_RECOMMENDATIONS_CSV_COLUMNS);
+      for (ApplyRecommendationResult result : applyRecommendationResults) {
+        AggregatedRecommendation aggregatedRecommendation =
+            aggregatedRecommendations.get(
+                RecommendationName.parse(result.getResourceName()).getRecommendationId());
+        Recommendation recommendation = aggregatedRecommendation.recommendation;
+        Campaign campaign = aggregatedRecommendation.campaign;
+
+        // The Recommendation.campaign field will be set only when the recommendation affects a
+        // single campaign. Otherwise, the query for recommendations will return 0 as the campaign
+        // ID. In that case, just set the 'NewOptiScore' column to 0.0 in the optimization reports.
+        csv.writeRow(
+            RecommendationName.parse(recommendation.getResourceName()).getRecommendationId(),
+            recommendation.getType().toString(),
+            aggregatedRecommendation.description,
+            String.valueOf(campaign.getId()),
+            String.valueOf(campaign.getOptimizationScore()),
+            String.valueOf(
+                campaign.getId() == 0
+                    ? 0.0
+                    : getCampaignOptiScore(googleAdsClient, customer.getId(), campaign.getId())));
+      }
+    }
+  }
+
+  /**
+   * Gets the optimization score for a customer.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID.
+   * @return the customer's optimization score.
+   */
+  private double getCustomerOptiScore(GoogleAdsClient googleAdsClient, long customerId) {
+    try (GoogleAdsServiceClient googleAdsServiceClient =
+        googleAdsClient.getLatestVersion().createGoogleAdsServiceClient()) {
+      return new AcquireOptimizations()
+          .getCustomerInfo(googleAdsServiceClient, customerId)
```

Comment:
This seems not needed to be an instance method. Do we have any location that we can place this method and share it between two files?